### PR TITLE
Implement fail-safe error handling for borg-extract 

### LINKED ISSUE #1231: Misleading error reporting in borg-extract
Hello darkness my old friend

```
            try:
                if dry_run:
                    archive.extract_item(item, dry_run=True)
                else:
                    if stat.S_ISDIR(item.mode):
                        dirs.append(item)
                        archive.extract_item(item, restore_attrs=False)
                    else:
                        archive.extract_item(item, stdout=stdout, sparse=sparse, hardlink_masters=hardlink_masters,
                                             original_path=orig_path)
            except OSError as e:
                self.print_warning('%s: %s', remove_surrogates(orig_path), e)
```

Guess we'll just rename input_io => something_else_io (say backup_io or ... or io_io!) and apply the same fix.

Note that this isn't nearly as critical as the other error handling bug, since nothing is written. So it's "merely" misleading error reporting.

### PR DESCRIPTION
**Pull request by enkore**
Note that this isn't nearly as critical as the other error handling bug,
since nothing is written. So it's "merely" misleading error reporting.

Fixes #1231

### CODE REVIEW

File: borg/archive.py [hunk 1]
```diff
@@ -356,54 +356,68 @@ def extract_item(self, item, restore_attrs=True, dry_run=False, stdout=False, sp
...
-            raise Exception('Unknown archive item type %r' % item[b'mode'])
+            elif stat.S_ISCHR(mode) or stat.S_ISBLK(mode):
+                os.mknod(path, item[b'mode'], item[b'rdev'])
+                self.restore_attrs(path, item)
+            else:
+                raise Exception('Unknown archive item type %r' % item[b'mode'])
 
     def restore_attrs(self, path, item, symlink=False, fd=None):
+        """
+        Restore filesystem attributes on *path* from *item* (*fd*).
```
Comment (enkore) [inline borg/archive.py]: `on *path* (*fd*) from *item*`

File: borg/archive.py [hunk 2]
```diff
@@ -59,18 +59,18 @@ def __str__(self):
...
+def backup_io():
+    """Context manager changing OSError to BackupOSError."""
     try:
         yield
     except OSError as os_error:
-        raise InputOSError(os_error) from os_error
+        raise BackupOSError(os_error) from os_error
 
 
 def input_io_iter(iterator):
```
Comment (ThomasWaldmann) [inline borg/archive.py]: backup_io_iter?

File: borg/archive.py [hunk 3]
```diff
@@ -46,7 +46,7 @@
 flags_noatime = flags_normal | getattr(os, 'O_NOATIME', 0)
 
 
-class InputOSError(Exception):
+class BackupOSError(Exception):
     """Wrapper for OSError raised while accessing input files."""
```
Comment (ThomasWaldmann) [inline borg/archive.py]: update the docstring also, please.

File: borg/archive.py (hunk 2 above)
Comment (enkore) [inline borg/archive.py]: Changed in the tests / forgot the production code. Ops.

**Comment by enkore** [comment]
live tested, ok. merge?

**Pull request merged**